### ISSUE TITLE
Plans for the future?

### ISSUE THREAD

**Issue by Friday9i** [lightvector/KataGo#10]
Sorry if opening an issue is not adequate, but I didn't see an alternative (ie a discord channel?).

First of all, huge congratulations, KataGo is an excellent program, very strong on 13x13 and 19x19, providing a lot of value added Go things: handicap, no weak play in end games, variable komi, variable board sizes, extremely fast and efficient improvement...
**That's just AMAZING, congratulations : -))).** LZ130 strength in 30 weeks x GPU, it's incredible: with current LZ computing power, that would take ~1 week where LZ took ~6 months!

Then a question: **what are the next steps?**
More precisely: 
- Is it possible to continue training it, and will you give us the opportunity to contribute (eg through an LZ approach, with an "autogtp" tool, including a windows version ; -)?
- What about larger nets (20, 30 or possibly 40 block)?
- Other ideas (over ones you already hinted: rectagular board sizes, japanese rules, ...)?

I'd be more than happy to switch my hardware from LZ and to dedicate it to train KataGo full time (on my Windows set-up)...

**Comment by lightvector** [comment]
Thanks for the enthusiasm and support!

KataGo isn't really set up right now to be run as a distributed project (for example, there's no master server that assigns self-play or match games), right now it's all one self-contained loop designed to be run on a set of machines with fast shared network filesystem they can read/write from. Although of course, there's no reason in theory why it couldn't be added, the whole thing would just need to be implemented - still a very nontrivial amount of coding work.

That doesn't mean there aren't plans and next steps! 

**Training process and making the neural net better**
Over the last month or two I've been continuing to test things at a slow but steady rate. For example, I've found that the learning rate, at least in the later stages of training, is set far too low, and multiplying it by a factor of 16 or 32 speeds up the neural net training considerably! I have some test nets that reach 100+ Elo higher than the original run with much less computation. Tentative estimates suggests that overall this might give a 1.5x or 2x boost in the rate of improvement, and solve some of the plateauing that was observed near the end of the current run. I'm also experimenting with LCB and KL-divergence ideas from LZ and LC0, which could improve things too. Also on the list is to compare squeeze-excite and other tweaks to neural net architecture.

I'll be doing a longer and better run some once I establish better hyperparameters and test all these new ideas.

**Major coding work** 
There are some things that aren't directly related to making the neural net stronger and require a lot of coding work, but would be very useful and valuable. None of these are super-easy and would require a dive into the guts of the code, but would be useful - I do plan to get to them myself eventually unless I find someone else interested in being a collaborator to help out.
  * For example, it would be very useful to have an OpenCL or other non-CUDA GPU implementation, as well as a CPU implementation of the neural net. I've designed for this, such that there's a clear way in the code where they could be added, but they aren't implemented yet. Having them would go a long way to making it easier to compile and run on a wider variety of systems.
  * Optimization and performance - KataGo is my first time ever writing GPU code or using something like CUDA. Also generally performance optimization isn't a thing I've worked on much. So there may be major possible improvements possible. For example, someone in discord chat noticed that LZ code using a 9x9 net that is actually larger than Kata's by a little bit gets more visits/s, at least on their hardware, so in theory it should be possible to make things faster than they are.
  * And yes, as you mentioned, rectangular boards, Japanese rules. I know how to do the first one, it just requires the work to sweep the code for all the places that only support squareness and generalize them (most of it actually already supports rectangle, you just have to fix the scattered spots that don't). Japanese rules requires some experimentation, but theoretically KataGo has a fully working implementation of Japanese rule for self-play, including all the quirks of Japanese scoring and rules about seki. And preliminary experiments showed that it worked, although I dropped it for the paper and the main run, since it does add a bunch of complexity to the rules. But it would not be hard to pick up and incorporate.

**User tools**
_This is probably the easiest category of things for an interested outside developer to contribute to._ As a lone dev with a separate full-time job, there's only so much I have time to do, so help is of course greatly appreciated. :)

KataGo has score and ownership predictions, which in theory should be useful for game reviews and human learning, or to use for auto-scoring game positions, or to analyze pro games or other stats (e.g. mistake frequency in points rather than only winrate, relation between handicap stones and komi, etc). But of course, KataGo is pretty new so of course there is not much support out there right now.

Extending existing tools or writing new tools using these things could be be incredibly cool and useful. Or at least I hope so. For example here (https://github.com/featurecat/lizzie/issues/505#issuecomment-481043303) I posted some one-off video visualizations I made (sadly, not a re-usable tool), which people seemed to like - I hope there's potential for things like this in the future. I would love to see someone take some of these outputs and make them into something real.

**Comment by Friday9i** [comment]
That's nice, thanks! That's so fun to play on alternative sizes (eg 13x13), with proper handicap and no slack moves when winning!
Once you have tested some more ideas, do you have any plan to merge the wonderful KataGo approach with the distributed training scheme of LZ? With a larger net (maybe 30b net?), that should be an absolute killer in no time ; -).

**Comment by poptangtwe** [comment]
Would the next run of KataGo handle Mirror Go better? 

KataGo is so excellent that the technique of KataGo may also benifit for other broad games like KataChess, KataXiangqi, KataShogi, KataGomuku.

**Comment by poptangtwe** [comment]
Would you consider implementing Fast Rollout witch used in AlphaGo Lee, AlphaGo Master, maybe also FineArt and Golaxy?
A member of the Chinese Leela competition team said, compared with Leela at 8×V100（20k playouts per second）, Golaxy and FineArt running at extremely high speeds in 2019“CHINA SECURITIES Cup” The World AI Weiqi Open.
Golaxy: more than 200k playouts per second on 16×V100
FineArt: several times more than simulations of Golaxy per second although we have no idea what hardware FineArt used.
@lightvector

**Comment by lightvector** [comment]
Closing old issue. For the record (if anyone stumbles across this page later) a *lot* of improvements have happened since this thread. Yay! Browsing back over the history of releases can give a good summary: https://github.com/lightvector/KataGo/releases
